src/devices/bus/msx/cart/ram.cpp

```diff
@@ -0,0 +1,349 @@
+		page(0)->install_ram(0x0000, 0x3fff, ram_data());
+		page(1)->install_ram(0x4000, 0x7fff, ram_data() + 0x4000);
+		page(2)->install_ram(0x8000, 0xbfff, ram_data() + 0x8000);
+	}
+};
+
+class msx_cart_64k_ram_device : public msx_cart_base_ram_device
+{
+public:
+	msx_cart_64k_ram_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+		: msx_cart_base_ram_device(mconfig, MSX_CART_64K_RAM, tag, owner, clock, 64 * 1024)
+	{ }
+
+protected:
+	virtual void device_start() override
+	{
+		msx_cart_base_ram_device::device_start();
+		page(0)->install_ram(0x0000, 0x3fff, ram_data());
+		page(1)->install_ram(0x4000, 0x7fff, ram_data() + 0x4000);
+		page(2)->install_ram(0x8000, 0xbfff, ram_data() + 0x8000);
+		page(3)->install_ram(0xc000, 0xffff, ram_data() + 0xc000);
+	}
+};
+
+
+class msx_cart_base_mm_ram_device : public msx_cart_base_ram_device
+{
+protected:
+	msx_cart_base_mm_ram_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock, u32 ram_size)
+		: msx_cart_base_ram_device(mconfig, type, tag, owner, clock, ram_size)
+		, m_rambank(*this, "rambank%u", 0U)
+		, m_bank_mask(0)
+	{ }
+
+	virtual void device_start() override;
+
+	template <int Bank> void bank_w(u8 data);
+
+private:
+	memory_bank_array_creator<4> m_rambank;
+	u8 m_bank_mask;
+};
+
+void msx_cart_base_mm_ram_device::device_start()
+{
+	msx_cart_base_ram_device::device_start();
+
+	m_bank_mask = device_generic_cart_interface::map_non_power_of_two(
+			unsigned(get_ram_size() / 0x4000),
+			[this] (unsigned entry, unsigned page)
+			{
+				for (int i = 0; i < 4; i++)
+					m_rambank[i]->configure_entry(entry, ram_data() + 0x4000 * page);
+			}
+	);
+
+	// The MSX system allows multiple devices to react to I/O, so we use taps.
+	io_space().install_write_tap(0xfc, 0xfc, "bank0", [this] (offs_t, u8& data, u8){ this->bank_w<0>(data); });
+	io_space().install_write_tap(0xfd, 0xfd, "bank1", [this] (offs_t, u8& data, u8){ this->bank_w<1>(data); });
+	io_space().install_write_tap(0xfe, 0xfe, "bank2", [this] (offs_t, u8& data, u8){ this->bank_w<2>(data); });
+	io_space().install_write_tap(0xff, 0xff, "bank3", [this] (offs_t, u8& data, u8){ this->bank_w<3>(data); });
+
+	// Not all memory mapper expansions allow reading of mapper registers.
+	io_space().install_read_tap(0xfc, 0xfc, "bank0", [this] (offs_t, u8& data, u8){ if (!machine().side_effects_disabled()) data = ~m_bank_mask | m_rambank[0]->entry(); });
+	io_space().install_read_tap(0xfd, 0xfd, "bank1", [this] (offs_t, u8& data, u8){ if (!machine().side_effects_disabled()) data = ~m_bank_mask | m_rambank[1]->entry(); });
+	io_space().install_read_tap(0xfe, 0xfe, "bank2", [this] (offs_t, u8& data, u8){ if (!machine().side_effects_disabled()) data = ~m_bank_mask | m_rambank[2]->entry(); });
+	io_space().install_read_tap(0xff, 0xff, "bank3", [this] (offs_t, u8& data, u8){ if (!machine().side_effects_disabled()) data = ~m_bank_mask | m_rambank[3]->entry(); });
```

Comment:
This always sets the data in the tap.  What actually happens when multiple devices react to the same I/O address?  Should it do a binary and instead or something?